[bugfix] Buffer will fail if freeing the last item

brianjjones opened the pull request:
Currently, Buffer will crash if it tries to free the last item,
because freeing it tries to look at the ->next of NULL. This adds
an extra check and also now returns once the handle has been freed,
since there's no reason to keep going.

grgustaf commented:
Not a bug,  *pItem was already tested in the while condition above. The if returns so you don't have to worry about that path.

brianjjones commented:
@grgustaf  this is only an issue if the last item is the one being freed.  The problem is that  *pItem = (*pItem)->next; causes *pItem to now be NULL.  The next line is pItem = &(*pItem)->next; which fails because NULL->next is not valid.  I updated this to only have the return once we have removed the handle.  Because it only happens if you remove the last element, and now we are returning, there's no need for the second check.

brianjjones commented:
In addition, there's no reason to keep iterating through the buffers once we free the handle, this saves time.

grgustaf commented:
OK, the return was missing. Yes, definitely correct now. +1
